# Dixième exercice en JavaScript

<img src="https://blog.univ-angers.fr/mathsinfo/files/2022/06/image-10.png">

*Résumé en français* : On vous donne une liste contenant des **couleurs** de **moufles** (donc pas de main gauche ou droite à distinguer). On vous demande le **nombre de paires** que vous pouvez **constituer**, c'est-à-dire avoir **2 moufles de la même couleur**. 

Avec le **premier exemple** donné, on peut constituer une **paire** de moufles **rouge** (red) **et** une paire **bleue** (blue) soit **2 paires**.

## Version classique

Relisez <a href="https://blog.univ-angers.fr/mathsinfo/2022/06/07/kata6/">l'exercice 6</a> que j'ai proposé, vous devriez constater de nombreuses ressemblances.

Une **première idée** est de commencer par **récupérer les différentes** couleurs puis, **pour chacune d'elle**, de compter **combien de fois** cette **couleur apparait**. Il suffira de **diviser par 2** ce nombre (sans tenir compte des virgules) pour savoir **combien de paires** on peut **constituer** avec cette couleur. Finalement, on fera la **somme** du **nombre de paires** trouvées.

Je reprends la **structure classique** vue dans l'exercice 6 :

In [1]:
var nb_paires = gants => 
{
    // On récupère les couleurs 
    couleurs = gants.reduce((a, g) => a.includes(g) ? a : a.concat(g), []);
    // Nb de paires trouvées
    paires = 0;
    for (c of couleurs)
    {
        // On compte le nombre de gants de la couleur c puis division par 2
        paires += Math.floor(gants.filter(g => g == c).length / 2)
    }
    return paires
}

In [2]:
nb_paires(["gray","black","purple","purple","gray","black"])

3

In [3]:
nb_paires(["red","green","blue"])

0

*Remarque* : La division entière par 2 ou des puissances de 2 (c’est-à-dire 4, 8, 16, 32…) peut se faire autrement. En effet, si un nombre est écrit en **binaire**, diviser par 2 revient à supprimer l’unité (**bit** le plus à droite) :

In [4]:
(13).toString(2)      // 13 en binaire

'1101'

In [5]:
(6).toString(2)       // Diviser par 2 = supprimer l'unité 

'110'

La raison est que 13 = <strong>1</strong> * 2^<span style="text-decoration: underline">3</span> + <strong>1</strong> * 2^<span style="text-decoration: underline">2</span> + <strong>0</strong> * 2^<span style="text-decoration: underline">1</span> + <strong>1</strong> = 1101<sub>2</sub>.<br>
En divisant par 2, toutes les puissances vont diminuer de 1 : 6 = <strong>1</strong> * 2^<span style="text-decoration: underline">2</span> + <strong>1</strong> * 2^<span style="text-decoration: underline">1</span> + <strong>0</strong> * 2^<span style="text-decoration: underline">0</span> = 110<sub>2</sub><br>

On utilise `>>` pour décaler un bit vers la droite

In [6]:
13 >> 1

6

De la même façon, on peut diviser par 4, multiplier par 8 etc :

In [7]:
13 >> 2        // Diviser par 4 (c'est-à-dire ÷ 2 fois par 2)

3

In [8]:
13 << 3        // Multiplier par 8 (c'est-à-dire × 3 fois par 2)

104

## Version ensembliste

Comme nous l'avons également vu dans l'exercice 6, nous pouvons utiliser des **ensembles** (`set`) pour récupérer les éléments **uniques**. Les versions sont alors nettement plus courtes :

In [9]:
var nb_paires = gants => [...new Set(gants)]
                .reduce((a,v) => a + (gants.filter(c => c == v).length >> 1), 0)

In [10]:
nb_paires(["gray","black","purple","purple","gray","black"])

3

## Version utilisant une seule boucle et une pile

Est-il vraiment nécessaire de récupérer les couleurs uniques ? Imaginons la situation dans la **vie réelle** avec `'red' 'green' 'red' 'blue' 'blue'` :

- Dans ma tête je pense à **0**, c'est le **nombre de paires** que j'ai réussi à faire
- Je prends le **premier** gant, il est `'red'`, je n'ai pas cette couleur donc je la garde
- Je prends le **deuxième**, il est `'green'`, je n'ai pas cette couleur donc je la garde
- Je prends le **troisième**, il est `'red'`. J'ai déjà cette couleur, je peux donc faire une **paire**. **J'enlève** les 2 moufles 'red' et **j'ajoute 1** au nombre de paires
- Je prends le **quatrième**, il est `'blue'`,  je n'ai pas cette couleur donc je la garde
- Je prends le **cinquième**, il est `'blue'`. J'ai déjà cette couleur, je peux donc faire une **paire**. J'enlève les 2 moufles 'blue' et **j'ajoute 1** au nombre de paires
- J'ai finalement réussi à constituer **2 paires**.

🤖 Refaites de tête le processus avec les gants `'red' 'red' 'red' 'red' 'red'`

**Prendre** un gant revient à **ajouter** un élément dans une liste, inversement **créer** une **paire** à **éliminer** cette couleur de la liste. Voyons comment ajouter, enlever ou tester si un élément est dans un ensemble en utilisant les méthodes `add`, `delete`, `has` :

In [11]:
paires = new Set()       // Ensemble vide
paires.add('red')        // on ajoute 'red' à l'ensemble
paires.add('blue')
paires

Set(2) { 'red', 'blue' }

In [12]:
paires.delete('red')     // on enlève 'red'
paires

Set(1) { 'blue' }

In [13]:
paires.has('red')        // on teste si 'red' est dans l'ensemble

false

Version finale basée sur cette idée :

In [14]:
var nb_paires = gants => {
  couleurs = new Set();
  paires = 0;
  for (c of gants)
  {
    if (couleurs.has(c)) 
    {   
      couleurs.delete(c);
      paires ++
    }
    else
      couleurs.add(c)
  }
  return paires
} 

In [15]:
nb_paires(["gray","black","purple","purple","gray","black"])

3